src/rpc/network/socket_communicator.cc

```diff
@@ -255,65 +288,107 @@ void SocketReceiver::Finalize() {
   }
   // Block main thread until all socket-threads finish their jobs
   for (auto& thread : threads_) {
-    thread.second->join();
+    thread->join();
   }
   // Clear all sockets
-  for (auto& socket : sockets_) {
-    socket.second->Close();
+  for (auto& group_sockets : sockets_) {
+    for (auto& socket : group_sockets) {
+      socket.second->Close();
+    }
   }
   server_socket_->Close();
   delete server_socket_;
 }
 
-void SocketReceiver::RecvLoop(TCPSocket* socket, MessageQueue* queue) {
-  CHECK_NOTNULL(socket);
-  CHECK_NOTNULL(queue);
-  for (;;) {
-    // If main thread had finished its job
-    if (queue->EmptyAndNoMoreAdd()) {
-      return;  // exit loop thread
-    }
-    // First recv the size
-    int64_t received_bytes = 0;
-    int64_t data_size = 0;
-    while (static_cast<size_t>(received_bytes) < sizeof(int64_t)) {
-      int64_t max_len = sizeof(int64_t) - received_bytes;
-      int64_t tmp = socket->Receive(
-        reinterpret_cast<char*>(&data_size)+received_bytes,
-        max_len);
-      CHECK_NE(tmp, -1);
-      received_bytes += tmp;
+int64_t RecvDataSize(TCPSocket* socket) {
+  int64_t received_bytes = 0;
+  int64_t data_size = 0;
+  while (static_cast<size_t>(received_bytes) < sizeof(int64_t)) {
+    int64_t max_len = sizeof(int64_t) - received_bytes;
+    int64_t tmp = socket->Receive(
+      reinterpret_cast<char*>(&data_size) + received_bytes,
+      max_len);
+    if (tmp == -1) {
+      if (received_bytes > 0) {
+          continue;
+      }
+      return -1;
     }
-    if (data_size < 0) {
-      LOG(FATAL) << "Recv data error (data_size: " << data_size << ")";
-    } else if (data_size == 0) {
-      // This is an end-signal sent by client
+    received_bytes += tmp;
+  }
+  return data_size;
+}
+
+void RecvData(TCPSocket* socket, char* buffer, const int64_t &data_size,
+  int64_t *received_bytes) {
+  while (*received_bytes < data_size) {
+    int64_t max_len = data_size - *received_bytes;
+    int64_t tmp = socket->Receive(buffer + *received_bytes, max_len);
+    if (tmp == -1) {
       return;
-    } else {
-      char* buffer = nullptr;
-      try {
-        buffer = new char[data_size];
-      } catch(const std::bad_alloc&) {
-        LOG(FATAL) << "Cannot allocate enough memory for message, "
-                   << "(message size: " << data_size << ")";
+    }
+    *received_bytes += tmp;
+  }
+}
+
+void SocketReceiver::RecvLoop(
+  std::unordered_map<int /* Sender (virtual) ID */,
+    std::shared_ptr<TCPSocket>> sockets,
+  std::unordered_map<int /* Sender (virtual) ID */,
+    std::shared_ptr<MessageQueue>> queues,
+  runtime::Semaphore *queue_sem) {
+  std::unordered_map<int, std::unique_ptr<RecvContext>> recv_contexts;
+  SocketPool socket_pool;
+  for (auto& socket : sockets) {
+    auto &sender_id = socket.first;
+    socket_pool.AddSocket(socket.second, sender_id);
+    recv_contexts[sender_id] = std::unique_ptr<RecvContext>(new RecvContext());
+  }
+
+  for (;;) {
```

Comment:
i don't quite understand the code here. maybe have some comments to explain it?

Reply:
done